rust/src/dcerpc/dcerpc.rs

```diff
@@ -0,0 +1,1794 @@
+
+    pub fn parse_request(&mut self, input: &[u8]) -> i32 {
+        let endianness = self.get_endianness();
+        match parser::dcerpc_parse_request(input, endianness) {
+            Ok((leftover_input, request)) => {
+                self.dcerpcrequest = Some(request);
+                println!("input len: {:?}, leftover_input: {:?}\n", input.len(), leftover_input.len());
+                let parsed = self.process_common_stub(&input, (input.len() - leftover_input.len()) as i32, core::STREAM_TOSERVER);
+                parsed
+            }
+            Err(nom::Err::Incomplete(_)) => {
+                // Insufficient data.
+                SCLogDebug!("Insufficient data while parsing DCERPC REQUEST");
+                -1
+            }
+            Err(_) => {
+                // Error, probably malformed data.
+                SCLogDebug!("An error occurred while parsing DCERPC REQUEST");
+                -1
+            }
+        }
+    }
+
+    pub fn handle_gap_ts(&mut self) -> u8 {
+        if self.buffer_ts.len() > 0 {
+            self.buffer_ts.clear();
+        }
+        return 0;
+    }
+
+    pub fn handle_gap_tc(&mut self) -> u8 {
+        if self.buffer_tc.len() > 0 {
+            self.buffer_tc.clear();
+        }
+        return 0;
+    }
+
+    pub fn dcerpc_parse(&mut self, input: &[u8], direction: u8) -> i32 {
+        let mut parsed = 0;
+        let retval;
+        let input_len = input.len();
+        let mut v: Vec<u8>;
+        let buffer = match direction {
+            core::STREAM_TOSERVER => {
+                if self.buffer_ts.len() + input_len > 1024 * 1024 {
+                    SCLogDebug!("To Server stream: Buffer Overflow");
+                    return -1;
+                }
+                v = self.buffer_ts.split_off(0);
+                v.extend_from_slice(input);
+                v.as_slice()
+            }
+            _ => {
+                if self.buffer_tc.len() + input_len > 1024 * 1024 {
+                    SCLogDebug!("To Client stream: Buffer Overflow");
+                    return -1;
+                }
+                v = self.buffer_tc.split_off(0);
+                v.extend_from_slice(input);
+                v.as_slice()
+            }
+        };
+        // Making it a hard requirement for a data stream to at least have the header in place
+        // else return -1 and the buffer stays empty
+        // Reason: header has the most important of data required for further proccessing
+        if self.bytes_consumed < 16 && input_len > 0 {
+            parsed = self.parse_dcerpc_header(&buffer);
+            if parsed == -1 {
+                return -1;
+            }
+            self.bytes_consumed += parsed as u16;
+        }
+
+        let fraglen = match self.get_hdr_fraglen() {
+            Some(x) => x,
+            None => 0,
+        };
+        if buffer.len() as u16 != fraglen {
+            println!("Possibly fragmented data, waiting for more..");
+            return 1;
```

Comment:
Fragmentation question: Is this the right way to handle fragmented data?
I figured there could be two ways to buffer the data:
1. when `Err::Incomplete` block is hit in the parsers
2. before any of the parsing begins (except header as we'd need to extract fraglen information from there)

I chose 2 and am waiting for all the data to come in and checking the buffer length against hdr.fraglen. Is this incorrect? What should be the right way to handle this?

Note to self: Add buffering logic to `Err::Incomplete` block of header.

Reply:
I think you should extend `self.buffer_tc` here with `buffer`.

However, I'm getting close to PoC to allow you to offload this task to the underlying API. So maybe wait a bit.

Reply:
See #4671

Reply:
This is now in `master`